x/transferinject/ibc_middleware.go

```diff
@@ -0,0 +1,193 @@
+	bankKeeper types.BankKeeper,
+) porttypes.ICS4Wrapper {
+	return &IBCMiddleware{
+		ICS4Wrapper:   ics,
+		rollappKeeper: rollappKeeper,
+		bankKeeper:    bankKeeper,
+	}
+}
+
+// NewIBCAckMiddleware creates a new IBCModule.
+// It intercepts acknowledged incoming IBC packets and adds token metadata that had just been registered on the rollapp itself,
+// to the local rollapp record.
+func NewIBCAckMiddleware(
+	ibc porttypes.IBCModule,
+	rollappKeeper types.RollappKeeper,
+) porttypes.IBCModule {
+	return &IBCMiddleware{
+		IBCModule:     ibc,
+		rollappKeeper: rollappKeeper,
+	}
+}
+
+// SendPacket wraps IBC ChannelKeeper's SendPacket function
+func (m *IBCMiddleware) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	destinationPort string, destinationChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if types.MemoAlreadyHasPacketMetadata(packet.Memo) {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnauthorized, "denom metadata already exists in memo")
+	}
+
+	rollapp, err := m.rollappKeeper.ExtractRollappFromChannel(ctx, destinationPort, destinationChannel)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrNotFound, "extract rollapp id from packet: %s", err.Error())
+	}
+	// TODO: consider that other chains also might want this feature
+	if rollapp == nil {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	if hasDenom(rollapp.TokenMetadata, packet.Denom) {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	denomMetadata, ok := m.bankKeeper.GetDenomMetaData(ctx, packet.Denom)
+	if !ok {
+		return 0, errorsmod.Wrapf(errortypes.ErrNotFound, "denom metadata not found")
+	}
+
+	packet.Memo = types.AddDenomMetadataToMemo(packet.Memo, denomMetadata)
+
+	data, err = types.ModuleCdc.MarshalJSON(packet)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrInvalidType, "marshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+}
+
+// OnAcknowledgementPacket adds the token metadata to the rollapp if it doesn't exist
+func (m *IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	var ack channeltypes.Acknowledgement
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return errorsmod.Wrapf(rtypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet acknowledgement: %v", err)
```

Comment:
fyi, there is an sdk error for failing to unmarshal json
but lets leave it for now